src/pie_models/models/components/pooler.py

```diff
@@ -0,0 +1,157 @@
+
+
+def pool_cls(hidden_state: Tensor, **kwargs) -> Tensor:
+    return hidden_state[:, 0, :]
+
+
+class AtIndexPooler(nn.Module):
+    """Pooler that takes the hidden state at a given index.
+
+    Args:
+        input_dim: The input dimension of the hidden state.
+        num_indices: The number of indices to pool.
+        offset: The offset to add to the indices.
+    """
+
+    def __init__(self, input_dim: int, num_indices: int = 2, offset: int = 0, **kwargs):
+        super().__init__(**kwargs)
+        self.input_dim = input_dim
+        self.num_indices = num_indices
+        self.offset = offset
+        self.missing_embeddings = nn.Parameter(torch.empty(num_indices, self.input_dim))
+        nn.init.normal_(self.missing_embeddings)
+
+    def forward(self, hidden_state: Tensor, indices: Tensor, **kwargs) -> Tensor:
+        batch_size, seq_len, hidden_size = hidden_state.shape
+        if indices.shape[1] != self.num_indices:
+            raise ValueError(
+                f"number of indices [{indices.shape[1]}] has to be the same as num_types [{self.num_indices}]."
+            )
+
+        # times num_types due to concat
+        result = torch.zeros(
+            batch_size, hidden_size * self.num_indices, device=hidden_state.device
+        )
+        for batch_idx, current_indices in enumerate(indices):
+            current_embeddings = [
+                hidden_state[batch_idx, current_indices[i] + self.offset, :]
+                if current_indices[i] >= 0
+                else self.missing_embeddings[i]
+                for i in range(self.num_indices)
+            ]
+            result[batch_idx] = cat(current_embeddings, 0)
+        return result
+
+    @property
+    def output_dim(self) -> int:
+        return self.input_dim * self.num_indices
+
+
+class ArgumentWrappedPooler(nn.Module):
+    """Wraps a pooler and maps the arguments to the pooler.
+
+    Args:
+        pooler: The pooler to wrap.
+        argument_mapping: A mapping from the arguments of the forward method to the arguments of the pooler.
+    """
+
+    def __init__(self, pooler: nn.Module, argument_mapping: Dict[str, str], **kwargs):
+        super().__init__(**kwargs)
+        self.pooler = pooler
+        self.argument_mapping = argument_mapping
+
+    def forward(self, hidden_state: Tensor, **kwargs) -> Tensor:
+        pooler_kwargs = {}
+        for k, v in kwargs.items():
+            if k in self.argument_mapping:
+                pooler_kwargs[self.argument_mapping[k]] = v
+        return self.pooler(hidden_state, **pooler_kwargs)
+
+    @property
+    def output_dim(self) -> int:
+        return self.pooler.output_dim
+
+
+class SpanMaxPooler(nn.Module):
+    """Pooler that takes the max hidden state over a span.
+
+    Args:
+        input_dim: The input dimension of the hidden state.
+        num_indices: The number of indices to pool.
+    """
+
+    def __init__(self, input_dim: int, num_indices: int = 2, **kwargs):
```

Comment:
You mean to shift the whole spans? Or individual offsets for start and end?

Reply:
a) consistency in method signature with AtIndexPooler. b) yes - why do we potentially want to shift indices, but not whole spans (I think you're explanation with 'shift because special tokens at start' holds here as well, does it not?)

Reply:
Shifting the indices with `AtIndexPooler` is only used for the `pooler_type == START_TOKENS` because in this case we want to have the token before the actual span (=the marker token) and do not want to implement a `AtOneBeforeIndexPooler` (which sounds a bit too special). Shifting with `SpanMaxPooler` for `offset=-1` would pool the span `[start-marker-token, mention-token_0, .., mention-token_-2]`, i.e. the last mention-token is missed out, so this seems not so useful in my opinion. In general, the preparation of any special version of indices should happen in the taskmodule. That we do it in the model for `pooler_type == START_TOKENS` is already a bit odd, but I think the case is simple enough to do that there.

However, I will have a look at this, if it does not significantly complicate the documentation and tests, I will include the offset for the `SpanMaxPooler`.  

PS: As usual, thanks a lot for your input :)

Reply:
> I think you're explanation with 'shift because special tokens at start' holds here as well, does it not?

After some more thinking: No, I think it doesn't. At least I can not come up with some useful example similar to the description of the offset parameter for `AtIndexPooler`. If you have any, I would be happy to integrate it.

Reply:
ok, now I understand. from the initial documentation, I thought offset was a general offset that could be induced on any sequence.